File: sdk/python/src/syfthub_sdk/_http.py
"""Internal HTTP client for SyftHub SDK."""

from __future__ import annotations

from typing import Any

import httpx

from syfthub_sdk.exceptions import (
    AccountingAccountExistsError,
    AccountingServiceUnavailableError,
    APIError,
    AuthenticationError,
    AuthorizationError,
    InvalidAccountingPasswordError,
    NotFoundError,
    ValidationError,
)
from syfthub_sdk.models import AuthTokens


class HTTPClient:
    """HTTP client with automatic token management."""

    def __init__(
        self,
        base_url: str,
        timeout: float = 30.0,
    ) -> None:
        """Initialize HTTP client.

        Args:
            base_url: Base URL for the API (e.g., "https://hub.syft.com")
            timeout: Request timeout in seconds
        """
        # Ensure base_url doesn't end with /
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout

        # Token storage
        self._access_token: str | None = None
        self._refresh_token: str | None = None

        # HTTP client
        self._client = httpx.Client(timeout=timeout)

    def close(self) -> None:
        """Close the HTTP client."""
        self._client.close()

    @property
    def is_authenticated(self) -> bool:
        """Check if client has tokens set."""
        return self._access_token is not None

    def set_tokens(self, tokens: AuthTokens) -> None:
        """Set authentication tokens.

        Args:
            tokens: AuthTokens with access_token and refresh_token
        """
        self._access_token = tokens.access_token
        self._refresh_token = tokens.refresh_token

    def get_tokens(self) -> AuthTokens | None:
        """Get current authentication tokens.

        Returns:
            AuthTokens if authenticated, None otherwise
        """
        if self._access_token and self._refresh_token:
            return AuthTokens(
                access_token=self._access_token,
                refresh_token=self._refresh_token,
            )
        return None

    def clear_tokens(self) -> None:
        """Clear authentication tokens."""
        self._access_token = None
        self._refresh_token = None

    def _get_headers(self, include_auth: bool = True) -> dict[str, str]:
        """Build request headers.

        Args:
            include_auth: Whether to include Authorization header

        Returns:
            Dictionary of headers
        """
        headers = {
            "Content-Type": "application/json",
            "Accept": "application/json",
        }
        if include_auth and self._access_token:
            headers["Authorization"] = f"Bearer {self._access_token}"
        return headers

    def _handle_error(self, response: httpx.Response) -> None:
        """Convert HTTP errors to SDK exceptions.

        Args:
            response: HTTP response object

        Raises:
            SyftHubError: Appropriate exception based on status code
        """
        status = response.status_code
        try:
            detail = response.json()
        except Exception:
            detail = response.text

        # Extract message from detail if it's a dict
        # Handle both {"detail": {...}} and {"detail": "string"} formats
        inner_detail = detail.get("detail", detail) if isinstance(detail, dict) else detail

        if isinstance(inner_detail, dict):
            message = inner_detail.get("message", str(inner_detail))
            error_code = inner_detail.get("code")
        else:
            message = str(inner_detail)
            error_code = None

        # Check for accounting-specific errors based on error code
        if isinstance(inner_detail, dict) and error_code:
            if error_code == "ACCOUNTING_ACCOUNT_EXISTS":
                raise AccountingAccountExistsError(message=message, detail=inner_detail)
            elif error_code == "INVALID_ACCOUNTING_PASSWORD":
                raise InvalidAccountingPasswordError(message=message, detail=inner_detail)
            elif error_code == "ACCOUNTING_SERVICE_UNAVAILABLE":
                raise AccountingServiceUnavailableError(message=message, detail=inner_detail)

        # Standard error handling based on status code
        if status == 401:
            raise AuthenticationError(message=message, detail=detail)
        elif status == 403:
            raise AuthorizationError(message=message, detail=detail)
        elif status == 404:
            raise NotFoundError(message=message, detail=detail)
        elif status == 422:
            raise ValidationError(message=message, detail=detail)
        else:
            raise APIError(
                message=message,
                status_code=status,
                detail=detail,
            )

    def _attempt_refresh(self) -> bool:
        """Attempt to refresh the access token.

        Returns:
            True if refresh succeeded, False otherwise
        """
        if not self._refresh_token:
            return False

        try:
            response = self._client.post(
                f"{self.base_url}/api/v1/auth/refresh",
                headers={"Content-Type": "application/json"},
                json={"refresh_token": self._refresh_token},
            )
            if response.status_code == 200:
                data = response.json()
                self._access_token = data["access_token"]
                self._refresh_token = data["refresh_token"]
                return True
        except Exception:
            pass
        return False

    def request(
        self,
        method: str,
        path: str,
        *,
        json: dict[str, Any] | None = None,
        params: dict[str, Any] | None = None,
        data: dict[str, Any] | None = None,
        include_auth: bool = True,
        retry_on_401: bool = True,
    ) -> dict[str, Any] | list[Any]:
        """Make an HTTP request.

        Args:
            method: HTTP method (GET, POST, PUT, PATCH, DELETE)
            path: URL path (will be joined with base_url)
            json: JSON body data
            params: Query parameters
            data: Form data (for login endpoint)
            include_auth: Whether to include Authorization header
            retry_on_401: Whether to retry with token refresh on 401

        Returns:
            Parsed JSON response

        Raises:
            SyftHubError: On API errors
        """
        url = f"{self.base_url}{path}"
        headers = self._get_headers(include_auth=include_auth)

        # For form data (OAuth2 login), adjust content type
        if data is not None:
            headers["Content-Type"] = "application/x-www-form-urlencoded"

        response = self._client.request(
            method=method,
            url=url,
            headers=headers,
            json=json,
            params=params,
            data=data,
        )

        # Handle 401 with token refresh
        if (
            response.status_code == 401
            and retry_on_401
            and include_auth
            and self._attempt_refresh()
        ):
            # Retry with new token
            return self.request(
                method=method,
                path=path,
                json=json,
                params=params,
                data=data,
                include_auth=include_auth,
                retry_on_401=False,  # Don't retry again
            )

        # Handle errors
        if response.status_code >= 400:
            self._handle_error(response)

        # Return parsed JSON (or empty dict for 204 No Content)
        if response.status_code == 204:
            return {}

        return response.json()  # type: ignore[no-any-return]

    def get(
        self,
        path: str,
        *,
        params: dict[str, Any] | None = None,
        include_auth: bool = True,
    ) -> dict[str, Any] | list[Any]:
        """Make a GET request."""
        return self.request("GET", path, params=params, include_auth=include_auth)

    def post(
        self,
        path: str,
        *,
        json: dict[str, Any] | None = None,
        data: dict[str, Any] | None = None,
        include_auth: bool = True,
    ) -> dict[str, Any] | list[Any]:
        """Make a POST request."""
        return self.request(
            "POST", path, json=json, data=data, include_auth=include_auth
        )

    def put(
        self,
        path: str,
        *,
        json: dict[str, Any] | None = None,
        include_auth: bool = True,
    ) -> dict[str, Any] | list[Any]:
        """Make a PUT request."""
        return self.request("PUT", path, json=json, include_auth=include_auth)

    def patch(
        self,
        path: str,
        *,
        json: dict[str, Any] | None = None,
        include_auth: bool = True,
    ) -> dict[str, Any] | list[Any]:
        """Make a PATCH request."""
        return self.request("PATCH", path, json=json, include_auth=include_auth)

    def delete(
        self,
        path: str,
        *,
        include_auth: bool = True,
    ) -> dict[str, Any] | list[Any]:
        """Make a DELETE request."""
        return self.request("DELETE", path, include_auth=include_auth)

File: sdk/typescript/src/http.ts
import {
  AccountingAccountExistsError,
  AccountingServiceUnavailableError,
  APIError,
  AuthenticationError,
  AuthorizationError,
  InvalidAccountingPasswordError,
  NetworkError,
  NotFoundError,
  ValidationError,
} from './errors.js';
import { toCamelCase, toSnakeCase, buildSearchParams } from './utils.js';

/**
 * Options for HTTP requests.
 */
export interface RequestOptions {
  /** Whether to include the Authorization header (default: true) */
  includeAuth?: boolean;
  /** Whether to send body as form-urlencoded instead of JSON */
  isFormData?: boolean;
  /** Request-specific timeout in milliseconds */
  timeout?: number;
}

/**
 * Internal request options including body and params.
 */
interface InternalRequestOptions extends RequestOptions {
  body?: unknown;
  params?: Record<string, unknown>;
}

/**
 * Auth tokens returned from login/refresh.
 */
export interface AuthTokens {
  accessToken: string;
  refreshToken: string;
  tokenType: string;
}

/**
 * Internal HTTP client for making API requests.
 *
 * Handles:
 * - Bearer token authentication
 * - Automatic token refresh on 401 responses
 * - JSON serialization/deserialization
 * - snake_case <-> camelCase conversion
 * - Error handling and exception mapping
 */
export class HTTPClient {
  private accessToken: string | null = null;
  private refreshToken: string | null = null;
  private isRefreshing = false;
  private refreshPromise: Promise<void> | null = null;

  /**
   * Create a new HTTP client.
   *
   * @param baseUrl - Base URL for all API requests (without trailing slash)
   * @param timeout - Default timeout in milliseconds (default: 30000)
   */
  constructor(
    private readonly baseUrl: string,
    private readonly timeout: number = 30000
  ) {}

  /**
   * Set authentication tokens.
   */
  setTokens(access: string, refresh: string): void {
    this.accessToken = access;
    this.refreshToken = refresh;
  }

  /**
   * Get current authentication tokens.
   */
  getTokens(): AuthTokens | null {
    if (!this.accessToken || !this.refreshToken) {
      return null;
    }
    return {
      accessToken: this.accessToken,
      refreshToken: this.refreshToken,
      tokenType: 'bearer',
    };
  }

  /**
   * Clear authentication tokens.
   */
  clearTokens(): void {
    this.accessToken = null;
    this.refreshToken = null;
  }

  /**
   * Check if the client has valid tokens.
   */
  hasTokens(): boolean {
    return this.accessToken !== null;
  }

  /**
   * Make a GET request.
   */
  async get<T>(
    path: string,
    params?: Record<string, unknown>,
    options?: RequestOptions
  ): Promise<T> {
    return this.request<T>('GET', path, { ...options, params });
  }

  /**
   * Make a POST request.
   */
  async post<T>(path: string, body?: unknown, options?: RequestOptions): Promise<T> {
    return this.request<T>('POST', path, { ...options, body });
  }

  /**
   * Make a PUT request.
   */
  async put<T>(path: string, body?: unknown, options?: RequestOptions): Promise<T> {
    return this.request<T>('PUT', path, { ...options, body });
  }

  /**
   * Make a PATCH request.
   */
  async patch<T>(path: string, body?: unknown, options?: RequestOptions): Promise<T> {
    return this.request<T>('PATCH', path, { ...options, body });
  }

  /**
   * Make a DELETE request.
   */
  async delete<T>(path: string, options?: RequestOptions): Promise<T> {
    return this.request<T>('DELETE', path, options);
  }

  /**
   * Make an HTTP request with automatic retry on 401.
   */
  private async request<T>(
    method: string,
    path: string,
    options: InternalRequestOptions = {}
  ): Promise<T> {
    const { includeAuth = true, isFormData = false, timeout, body, params } = options;

    // Build URL with query params
    let url = `${this.baseUrl}${path}`;
    if (params) {
      const searchParams = buildSearchParams(params);
      const queryString = searchParams.toString();
      if (queryString) {
        url += `?${queryString}`;
      }
    }

    // Build headers
    const headers: Record<string, string> = {};

    if (includeAuth && this.accessToken) {
      headers['Authorization'] = `Bearer ${this.accessToken}`;
    }

    // Build body
    let requestBody: string | undefined;
    if (body !== undefined) {
      if (isFormData) {
        headers['Content-Type'] = 'application/x-www-form-urlencoded';
        // For form data, convert to URLSearchParams
        const formData = new URLSearchParams();
        for (const [key, value] of Object.entries(body as Record<string, unknown>)) {
          if (value !== undefined && value !== null) {
            formData.append(key, String(value));
          }
        }
        requestBody = formData.toString();
      } else {
        headers['Content-Type'] = 'application/json';
        // Convert camelCase to snake_case for JSON bodies
        requestBody = JSON.stringify(toSnakeCase(body));
      }
    }

    // Create abort controller for timeout
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout ?? this.timeout);

    try {
      const response = await fetch(url, {
        method,
        headers,
        body: requestBody,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      // Handle 401 with automatic token refresh
      if (response.status === 401 && includeAuth && this.refreshToken) {
        // Attempt to refresh the token
        await this.attemptTokenRefresh();

        // Retry the original request with new token
        return this.request<T>(method, path, {
          ...options,
          // Mark that we shouldn't retry again to prevent infinite loops
          includeAuth: true,
        });
      }

      return await this.handleResponse<T>(response);
    } catch (error) {
      clearTimeout(timeoutId);

      if (error instanceof SyftHubError) {
        throw error;
      }

      if (error instanceof Error) {
        if (error.name === 'AbortError') {
          throw new NetworkError('Request timed out', error);
        }
        throw new NetworkError(error.message, error);
      }

      throw new NetworkError('Unknown network error');
    }
  }

  /**
   * Handle the HTTP response and convert to the expected type.
   */
  private async handleResponse<T>(response: Response): Promise<T> {
    // Try to parse response as JSON
    let data: unknown;
    const contentType = response.headers.get('content-type');

    if (contentType?.includes('application/json')) {
      try {
        data = await response.json();
      } catch {
        data = null;
      }
    } else {
      // For non-JSON responses, use text
      const text = await response.text();
      data = text || null;
    }

    // Handle error responses
    if (!response.ok) {
      this.handleErrorResponse(response.status, data);
    }

    // Convert snake_case to camelCase and parse dates
    return toCamelCase<T>(data);
  }

  /**
   * Handle error responses by throwing appropriate exceptions.
   */
  private handleErrorResponse(status: number, data: unknown): never {
    const message = this.extractErrorMessage(data);
    const { code, detail } = this.extractErrorCodeAndDetail(data);

    // Check for accounting-specific errors based on error code first
    if (code) {
      switch (code) {
        case 'ACCOUNTING_ACCOUNT_EXISTS':
          throw new AccountingAccountExistsError(message, detail);
        case 'INVALID_ACCOUNTING_PASSWORD':
          throw new InvalidAccountingPasswordError(message, detail);
        case 'ACCOUNTING_SERVICE_UNAVAILABLE':
          throw new AccountingServiceUnavailableError(message, detail);
      }
    }

    // Standard status code handling
    switch (status) {
      case 401:
        throw new AuthenticationError(message);
      case 403:
        throw new AuthorizationError(message);
      case 404:
        throw new NotFoundError(message);
      case 422:
        throw new ValidationError(message, this.extractValidationErrors(data));
      default:
        throw new APIError(message, status, data);
    }
  }

  /**
   * Extract error code and detail from API response.
   * Used for accounting-specific error handling.
   */
  private extractErrorCodeAndDetail(
    data: unknown
  ): { code?: string; detail?: unknown } {
    if (!data || typeof data !== 'object') {
      return {};
    }

    // FastAPI returns { detail: { code: "...", message: "...", ... } }
    if ('detail' in data) {
      const detail = (data as { detail: unknown }).detail;
      if (detail && typeof detail === 'object' && 'code' in detail) {
        const innerDetail = detail as { code?: string };
        return {
          code: innerDetail.code,
          detail: detail,
        };
      }
    }

    return {};
  }

  /**
   * Extract error message from API response.
   */
  private extractErrorMessage(data: unknown): string {
    if (typeof data === 'string') {
      return data;
    }

    if (data && typeof data === 'object') {
      // FastAPI style: { detail: "message" }
      if ('detail' in data) {
        const detail = (data as { detail: unknown }).detail;
        if (typeof detail === 'string') {
          return detail;
        }
        // FastAPI validation errors: { detail: [{ msg: "...", loc: [...] }] }
        if (Array.isArray(detail) && detail.length > 0) {
          const firstError = detail[0] as { msg?: string };
          if (firstError?.msg) {
            return firstError.msg;
          }
        }
      }

      // Generic: { message: "..." } or { error: "..." }
      if ('message' in data && typeof (data as { message: unknown }).message === 'string') {
        return (data as { message: string }).message;
      }
      if ('error' in data && typeof (data as { error: unknown }).error === 'string') {
        return (data as { error: string }).error;
      }
    }

    return 'An error occurred';
  }

  /**
   * Extract field-level validation errors from API response.
   */
  private extractValidationErrors(data: unknown): Record<string, string[]> | undefined {
    if (!data || typeof data !== 'object' || !('detail' in data)) {
      return undefined;
    }

    const detail = (data as { detail: unknown }).detail;
    if (!Array.isArray(detail)) {
      return undefined;
    }

    const errors: Record<string, string[]> = {};

    for (const error of detail) {
      if (typeof error === 'object' && error !== null && 'loc' in error && 'msg' in error) {
        const { loc, msg } = error as { loc: unknown[]; msg: string };
        // loc is typically ['body', 'field_name']
        const field = String(loc[loc.length - 1] ?? 'unknown');
        if (!errors[field]) {
          errors[field] = [];
        }
        errors[field].push(msg);
      }
    }

    return Object.keys(errors).length > 0 ? errors : undefined;
  }

  /**
   * Attempt to refresh the access token using the refresh token.
   */
  private async attemptTokenRefresh(): Promise<void> {
    // If already refreshing, wait for the existing refresh to complete
    if (this.isRefreshing && this.refreshPromise) {
      await this.refreshPromise;
      return;
    }

    this.isRefreshing = true;

    this.refreshPromise = (async () => {
      try {
        const response = await fetch(`${this.baseUrl}/api/v1/auth/refresh`, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({ refresh_token: this.refreshToken }),
        });

        if (!response.ok) {
          // Refresh failed, clear tokens
          this.clearTokens();
          throw new AuthenticationError('Token refresh failed');
        }

        const data = (await response.json()) as {
          access_token: string;
          refresh_token: string;
        };

        this.accessToken = data.access_token;
        this.refreshToken = data.refresh_token;
      } finally {
        this.isRefreshing = false;
        this.refreshPromise = null;
      }
    })();

    await this.refreshPromise;
  }
}

// Import SyftHubError for type checking
import { SyftHubError } from './errors.js';
